isoutils/isoutils.go

```diff
@@ -0,0 +1,517 @@
+)
+
+type Paths int
+
+const (
+	clr_efi     Paths = iota + 0
+	clr_img_efi       // Location to mount the EFI partition in the passed-in img file
+	clr_initrd
+	clr_rootfs
+	clr_cdroot
+)
+
+var (
+	tmpPaths = make([]string, 5)
+)
+
+func mkTmpDirs() error {
+	msg := "Creating directory trees"
+	prg := progress.NewLoop(msg)
+	log.Info(msg)
+	var err error
+
+	tmpPaths[clr_efi], err = ioutil.TempDir("", "clr_efi-")
+	if err != nil {
+		prg.Failure()
+		return (err)
+	}
+	tmpPaths[clr_initrd], err = ioutil.TempDir("", "clr_initrd-")
+	if err != nil {
+		prg.Failure()
+		return (err)
+	}
+	tmpPaths[clr_cdroot], err = ioutil.TempDir("", "clr_cdroot-")
+	if err != nil {
+		prg.Failure()
+		return (err)
+	}
+
+	/* Create specific directories for the new cd's root */
+	for _, d := range []string{
+		tmpPaths[clr_cdroot] + "/isolinux",
+		tmpPaths[clr_cdroot] + "/EFI",
+		tmpPaths[clr_cdroot] + "/images",
+		tmpPaths[clr_cdroot] + "/kernel",
+	} {
+		if _, err := os.Stat(d); os.IsNotExist(err) {
+			os.Mkdir(d, os.ModePerm)
+		}
+	}
+
+	prg.Success()
+	return err
+}
+
+func mkRootfs() error {
+	msg := "making rootfs squashfs"
+	prg := progress.NewLoop(msg)
+	log.Info(msg)
+
+	/* TODO: This takes a long time to run, it'd be nice to see it's output as it's running */
+	args := []string{
+		"mksquashfs",
+		tmpPaths[clr_rootfs],
+		tmpPaths[clr_cdroot] + "/images/rootfs.img",
+		"-comp",
+		"lzo",
+		"-wildcards",
+		"-ef",
+		"/home/blovin/go/src/github.com/clearlinux/clr-installer/isoutils/exclude", //TODO: Fix hardcoded value
+	}
+	err := cmd.RunAndLog(args...)
+	if err != nil {
+		prg.Failure()
+		return err
+	}
+	prg.Success()
+	return err
+}
+
+func mkInitrd(version string, model *model.SystemInstall) error {
+	msg := "Installing the base system for initrd"
+	prg := progress.NewLoop(msg)
+	log.Info(msg)
+
+	var err error
+	options := args.Args{
+		SwupdMirror:             model.SwupdMirror,
+		SwupdStateDir:           tmpPaths[clr_initrd] + "/var/lib/swupd/",
+		SwupdStateClean:         true,
+		SwupdFormat:             "staging",
+		SwupdSkipDiskSpaceCheck: true,
+	}
+	sw := swupd.New(tmpPaths[clr_initrd], options)
+
+	/* Should install the overridden CoreBundles above (eg. os-core only) */
+	if err := sw.Verify(version, model.SwupdMirror, true); err != nil {
+		prg.Failure()
+		return err
+	}
+
+	prg.Success()
+	return err
+}
+
+func mkInitrdInitScript(templatePath string) error {
+	msg := "Creating and installing init script to initrd"
+	prg := progress.NewLoop(msg)
+	log.Info(msg)
+
+	type Modules struct {
+		Modules []string
+	}
+	mods := Modules{}
+
+	//Modules to insmod during init, paths relative to the kernel folder
+	modules := []string{
+		"/kernel/fs/isofs/isofs.ko",
+		"/kernel/drivers/cdrom/cdrom.ko",
+		"/kernel/drivers/scsi/sr_mod.ko",
+		"/kernel/fs/overlayfs/overlay.ko",
+	}
+
+	/* Find kernel, then break the name into kernelVersion */
+	kernelGlob, err := filepath.Glob(tmpPaths[clr_rootfs] + "/lib/kernel/org.clearlinux.native.*")
```

Comment:
We will need to update this somewhat because we might not always have the native kernel; could be kvm or lts or ...:
```
# ls /lib/kernel/org.clearlinux.*
/lib/kernel/org.clearlinux.kvm.5.0.3-315
/lib/kernel/org.clearlinux.lts2018.4.19.30-15
/lib/kernel/org.clearlinux.native.5.0.3-721
```

Reply:
Fixed, this is now generic and doesn't rely on native - I tested with kvm.yaml and can produce an ISO now.